BaragonService/src/main/java/com/hubspot/baragon/service/elb/ApplicationLoadBalancer.java

```diff
@@ -29,23 +62,401 @@ public ApplicationLoadBalancer(Optional<ElbConfiguration> configuration,
+    String vpcId = targetGroup.getVpcId();
+
+    for (BaragonAgentMetadata agent : agents) {
+      if (agent.getEc2().getVpcId().isPresent()) {
+        if (! agent.getEc2().getVpcId().get().equals(vpcId)) {
+          LOG.warn("Agent {} not on same VPC as its target group {}", agent, targetGroup);
+        }
+      } else {
+        LOG.warn("Agent {} not assigned to a VPC", agent);
+      }
+    }
+
+    for (LoadBalancer loadBalancer : loadBalancers) {
+      if (!vpcId.equals(loadBalancer.getVpcId())) {
+        LOG.warn("Load balancer {} on different VPC from target group {}", loadBalancer, targetGroup);
+      }
+    }
+  }
+
+  private void guaranteeAzEnabled(Collection<BaragonAgentMetadata> agents,
+                                  Collection<LoadBalancer> loadBalancers) {
+    for (LoadBalancer loadBalancer : loadBalancers) {
+      Collection<String> azNames = new HashSet<>();
+      for (AvailabilityZone availabilityZone : loadBalancer.getAvailabilityZones()) {
+        azNames.add(availabilityZone.getZoneName());
+      }
+
+      for (BaragonAgentMetadata agent : agents) {
+        if (agent.getEc2().getAvailabilityZone().isPresent()
+            && ! azNames.contains(agent.getEc2().getAvailabilityZone().get())) {
+          guaranteeHasAllSubnets(
+              agent.getEc2().getSubnetId().asSet(),
+              loadBalancer);
+        }
+      }
+    }
+  }
+
+  /**
+   *
+   * @param target Target to check
+   * @param targetGroup Group to check in
+   * @return if the given target is the last healthy target in the given target group
+   */
+  private boolean isLastHealthyInstance(TargetDescription target, TargetGroup targetGroup) {
+    DescribeTargetHealthRequest targetHealthRequest = new DescribeTargetHealthRequest()
+        .withTargetGroupArn(targetGroup.getTargetGroupArn());
+    List<TargetHealthDescription> targetHealthDescriptions = elbClient
+        .describeTargetHealth(targetHealthRequest)
+        .getTargetHealthDescriptions();
+
+    boolean instanceIsHealthy = false;
+    int healthyCount = 0;
+
+    for (TargetHealthDescription targetHealthDescription : targetHealthDescriptions) {
+      if (targetHealthDescription.getTargetHealth().getState()
+          .equals(TargetHealthStateEnum.Healthy.toString())) {
+        healthyCount += 1;
+        if (targetHealthDescription.getTarget().equals(target)) {
+          instanceIsHealthy = true;
+        }
+      }
+    }
+
+    return instanceIsHealthy && healthyCount == 1;
+  }
+
+  private void guaranteeHasAllSubnets(Collection<String> subnetIds, LoadBalancer loadBalancer) {
+    Collection<String> subnetsOnLoadBalancer = getSubnetsFromLoadBalancer(loadBalancer);
+    Collection<String> subnetsToAdd = new HashSet<>();
+    for (String subnetId : subnetIds) {
+      if (! subnetsOnLoadBalancer.contains(subnetId)) {
+        subnetsToAdd.add(subnetId);
+      }
+    }
+
+    try {
+      SetSubnetsRequest subnetsRequest = new SetSubnetsRequest()
```

Comment:
does SetSubnets override all current ones? If so we'd want to make sure this list also includes ones currently active so we aren't removing any that might still be needed (i.e. subnetsOnLoadBalancer + subnetIds here)